Add logging docs (#754)
appsignal/appsignal-docs: source/posts/application/logging.mdx

---
title: "Logging"
sup: "BETA"
---

<Notice>
  🧪 Logs are currently only available to Beta Testers. If you think your
  application is a good candidate, [send us an
  email](mailto:support@appsignal.com?subject=Logging%20beta)!
</Notice>

With AppSignal's Logs feature, you can store, view, and query your applications logs,
giving you extra insight into your application's performance and behavior.

This guide explains how to configure your application to send logs to AppSignal,
and how you can interact with those logs from within the AppSignal app.

## Setting up Logs

<Notice>

>   ⚠️ AppSignal currently only supports SysLog for configuring your application's

Reviewer: ```suggestion
  ⚠️ AppSignal currently only supports Syslog for configuring your application's
```
Is this proper spelling?

  logging.
</Notice>

### Creating a Log Source

> To send logs to AppSignal, you'll first need to configure your log sources. You will need to do this once per log source.

Author: Note to self to come back to this.

#### Choose The Log Format

Next you will need to let AppSignal know how your applications logs will be formatted. AppSignal supports three types of log formatting:

- [plaintext](#plaintext)
- [logfmt](#logfmt)
- [json](#json)

When using [logfmt](#logfmt) or [json](#json) as your log format, you can add searchable attributes to your logs. Attributes can be any of the following types:

- string
- integer (e.g. 10)
- double (e.g. 10.01)
- boolean (true/false)

##### plaintext

When specifying plaintext as your log format, AppSignal will not parse data it receives from your source.

##### logfmt

<Notice>
  📖 **logfmt** (fmt stands for format) is a logging convention developed by
  Heroku. You can read more about logfmt [here](https://brandur.org/logfmt).
</Notice>

When specifying logfmt as your log format, AppSignal will parse logs received from your source based on key-value pairs, storing each key and value as attributes on the log line, allowing you to filter log lines on these attributes.

**Key behavior:**

- If a key does not have a value, it will be appended to the message field.

> - If a key-value pair has a key with the value of `msg` or `message` the value of this key will be parsed as the log message.

Reviewer 2: ```suggestion
- If a key-value pair has a key with the name of `msg` or `message` the value of this key will be parsed as the log message.
```

- If no message/msg key is available, the entire log line will be considered the message.

###### Examples

1. `this is foo=bar duration=10 a value="with spaces" message` <br/>will be parsed to a log line with the message `"this is a message"`, and the following attributes:<br/>`{ "foo": "bar", "duration": 10, "value": "with spaces"}`

1. `this is foo=bar a duration=10 message message="I am a message"` <br/>will be parsed to a log line with the message `"I am a message"`, and the following attributes: <br/>`{ "foo": "bar", "duration": 10 }`

1. `foo=bar duration=100` <br/>will be parsed to a log line with the message `"foo=bar duration=100"`, and the following attributes: <br/> `{ "foo": "bar", "duration": 10 }`

##### JSON

When specifying json as your log format, AppSignal will parse logs received from your source based on key-value pairs, storing each key and value as attributes on the log line, allowing you to filter log lines on these attributes.

**Key behavior:**

> - If a key-value pair has a key with the value of `msg` or `message` the value of this key will be parsed as the log message.

Reviewer 2: ```suggestion
- If a key-value pair has a key with the name of `msg` or `message` the value of this key will be parsed as the log message.
```

- If no message/msg key is available, the entire log line will be considered the message.

###### Examples

1. `{"msg": "this is a message", duration: 10, "foo": "bar"}`<br />

   will be parsed to a log line with the message `"this is a message"`, and the following attributes: <br />
   `{ "foo": "bar", "duration": 10, "value": "with spaces"}`

1. `{"foo": "bar", "duration": 10}`<br />

   will be parsed to a log line with the message `"{"foo": "bar", "duration": 10}"`, and the following attributes:<br />

   `{ "foo": "bar", "duration": 10}`

#### Completing Setup

Once your new source is created, it will be listed in the Log Sources table on the Logs page, alongside a generated API key.

![Screenshot of Log Sources table highlighting API key](/assets/images/screenshots/logging/highlight-api.png)

This API key is important, it tells AppSignal where to catalog the logs you send it.

<Notice>
  ⏱ As AppSignal only stores logs that are sent to it from your application, you
  will only be able to see logs created **after** you've completed your
  configuration.
</Notice>

#### Syslog Configuration

> ##### Generating a new SysLog Config File

Reviewer: ```suggestion
##### Generating a new Syslog Config File
```

Syslog is a standard protocol for message logging. Syslog can read log files or receive Syslog messages directly from applications such as Nginx and write those logs to a file or send them to a third-party application.

Use the code snippet below to configure `rsyslog` for `SSL log` forwarding with the API key for this source.

```
1. # /etc/rsyslog.d/appsignal.conf
2. # Use SSL for TCP syslog
3. $ActionSendStreamDriver gtls
4. $ActionSendStreamDriverMode 1
5. $ActionSendStreamDriverAuthMode anon
6. $DefaultNetstreamDriverCAFile /etc/ssl/certs/ca-certificates.crt

> 7. $template AppsignalFormat,"<%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% %procid% %msgid% [ls-86f1d33e-df8f-4e74-b39d-3f23971c0983@59115] %msg%\n"

Reviewer 3: That's part of the API key. It's not clear at all that you need to replace `ls-86f1d33e-df8f-4e74-b39d-3f23971c0983` (but not `@59115` right after that!) with the API key for your logging source. Luckily, the app does give you the already-interpolated snippet.

> 7. $template AppsignalFormat,"<%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% %procid% %msgid% [ls-86f1d33e-df8f-4e74-b39d-3f23971c0983@59115] %msg%\n"

Author: We can omit this section, it no longer makes sense - earlier there was no installation instruction, and it was unsure how this would be done - but with the current product having this installation section, the docs just need to explain this adequately instead :) thanks!

> 8. @@appsignal-endpoint.net:6514;AppsignalFormat

Reviewer 2: Do those line numbers (?) need to be there? If not, it makes copy-pasting this file content confusing because people need to remove those line numbers. I'd leave them out.

```

To only log specific lines you can do so like in the example below, where only ngnix lines are logged. Simply modify the codesnippet to suit your applications needs and include it in your `appsignal.conf` file.

`if $programname == 'nginx' then @@appsignal-endpoint.net:6514;AppsignalFormat`

##### Saving Your Config

We advise keeping the filename, and sourcename identical.

If you are unsure about this step you can always [reach out](/support) for assistance. We'll help you find the answer!

> Once you've updated the relevant config file, your Logs should become visible to AppSignal. You'll be able to view and query your logs from the AppSignal application.

Reviewer 2: They don't need to restart or reload config anywhere?

## Log Management

![Screenshot of expanded log](/assets/images/screenshots/logging/all-logs.png)

Once succesfully configured, you'll be able to see logs for all of your sources from within the AppSignal application. You can filter, and query logs and expand specific logs for their trace data. When viewing your logs you can:

- [Filter and Query Logs](#filter-logs)
- [Copy Log Meta Data As JSON](#Copy-Log-Meta-Data-As-JSON)
- [Share Log](#Share-Log)
- [View What Happened Here](#What-Happened-Here)
- [Time Travel](#Time-Travel)
- [View Live Trail](#Live-Trail)

#### Filter and Query Logs

Use the `Source ˬ` drop down to filter which sources you see the logs for.

> Use the `Severity ˬ` drop down to filter the severity of logs

Reviewer 2: That dropdown symbol renders a bit weird.
![image](https://user-images.githubusercontent.com/282402/191038711-a68eb96d-6eff-4e81-b6f2-431533d22f21.png)

Author: Yeah, I will request a feature for more accurate button rendering, I wanted to see how replicable things where within the current constraints of mdx.

Filtering and Querying is further explained in the [Filtering and Querying](#Filtering-and-Querying) section of this guide.

#### Copy Log Meta Data As JSON

Expand a log by pressing the `+` button, Click `Copy as JSON`. Meta data will automatically be copied to your clipboard as a valid JSON object.

#### Share Log

Click `🔗 Copy link to log line` and a link to this log line will automatically be copied to your clipboard.

<Notice>
  Only users within your AppSignal Organisation will be able to access the logs
  via this link
</Notice>

#### What Happened Here

Want to see a snapshot of your applications metric at the exact time the logline was made? Click `⌖ What Happened Here`

#### Jump to Date

> Something happen while you were having coffee? No problem. Use the `Jump to date ˬ` button to navigate to the date and time you want to investigate.

Reviewer 3: ```suggestion
Something happened while you were having coffee? No problem. Use the `Jump to date ˬ` button to navigate to the date and time you want to investigate.
```

#### Live Trail

Clicking `Go to live ▶` will take you to a live trail of your logs.

## Filtering and Querying

AppSignal supports log filtering by:

> | Filter Field | Description                                                                    |
> | ------------ | ------------------------------------------------------------------------------ |
> | `Source`     | A Log Source from your application, eg MongoDB or Sidekiq                      |
> | `Severity`   | Log severity level. Accepted values are: error, warning, info, debug and trace |
> | `Hostname`   | Application Host (via search)                                                  |
> | `Group`      | Namespaces defined for your application (via search)                           |
> | `Message`    | Log message contents (via search)                                              |
> | `Attribute`  | Attributes defined in your logs (via search)                                   |

Reviewer 3: The filter fields shouldn't be capitalised here, since they are later shown in the search syntax examples without capitalising, which is how they need to be, I think.

```suggestion
| Filter Field | Description                                                                    |
| ------------ | ------------------------------------------------------------------------------ |
| `source`     | A Log Source from your application, eg MongoDB or Sidekiq                      |
| `severity`   | Log severity level. Accepted values are: error, warning, info, debug and trace |
| `hostname`   | Application Host (via search)                                                  |
| `group`      | Namespaces defined for your application (via search)                           |
| `message`    | Log message contents (via search)                                              |
| `attribute`  | Attributes defined in your logs (via search)                                   |
```

### Search Syntax Explained

The below table explains the current AppSignal Logs seach query syntax.

> Queries can be combined with a space between them (see Combined Queryes in [Search Query Overview](#Search-Query-Overview)).

Reviewer 3: typo
```suggestion
Queries can be combined with a space between them (see Combined Queries in [Search Query Overview](#Search-Query-Overview)).
```

Author: Queryes sounds like a GOT character

| Less Than                      | `attributes.duration < 100`                                                 | Returns all logs where `duration` is less than 100                                                                                                                                      |
| Combined Queries               | `banana group="slow queries" "not found" message!:segfault group!=frontend` | Returns all logs where logs contain a value like `Banana` from the group `slow queries` where log is like `not found` and where log is not like `segfault` and group is not `frontend`. |

## Managing Sources

To modify a source go to AppSignal > Logging > Sources and click `Manage source`

![Screenshot of Source Management](/assets/images/screenshots/logging/view-source.png)

When managing a source, you can:

- [Edit Source](#Edit-Source)
- [Remove Source](#remove-source)

### Edit Source

To edit the source click `Edit Source` in the top right corner of the screen. You can edit the following source attributes:

- Name
- Platform (currently limited to syslog)
- Format

> Any changes made (excluding name) will not be applied retrospectively, and therefor will only apply to new logs.

Reviewer: ```suggestion
Any changes made (excluding name) will not be applied retrospectively, and therefore will only apply to new logs.
```

therefor sounds a bit too archaic

Reviewer 3: also
```suggestion
Any changes made (excluding name) will not be applied retroactively, and therefor will only apply to new logs.
```

Author: Therefor and Therefore have quite different meanings, [Grammarly can explain](https://www.grammarly.com/blog/therefore-vs-therefor/#:~:text=Therefor,-Marko%20Ticak&text=Therefore%20is%20an%20adverb%20that,%E2%80%9D%20or%20%E2%80%9Cfor%20it.%E2%80%9D)

### Remove Source

Deletion is permanent and will revoke the API key used for that source.

> Before removing a log source, we recommend ensuring that the source is no longer sending logs to AppSignal to ensure your application is not making unnecessary calls to third parties.

Reviewer 3: _How?_ (this is an example, not sure if rebooting rsyslog is necessary)
```suggestion
Before removing a log source, we recommend ensuring that the source is no longer sending logs to AppSignal so that your application is not making unnecessary calls to third parties. Before removing the log source in AppSignal, remove the rsyslog configuration file for this log source and restart rsyslog.
```

![Screenshot of confirm deletion alert](/assets/images/screenshots/logging/delete-source-confirmation.png)

To remove a source click `Remove source` from the Source Management page. When prompted to confirm removal read the dialogue box carefully and click `Yes, delete this`. The source will be deleted.

> ## Trouble Shooting

Reviewer: ```suggestion
## Troubleshooting
```

If your having trouble with rsyslog you may wish to follow some of the troubleshooting information in this documentation. Alternatively, you can review the official [Rsyslog Troubleshooting Documentation](https://www.rsyslog.com/doc/master/troubleshooting/index.html). If that still doesn't solve your problems, you can [contact us](/support), and we'll help you figure things out!

### Check Installation

To see if syslog is installed, you can run the following terminal command:

> `ps -ef |grep iterm`

Reviewer 2: ```suggestion
`ps -ef | grep iterm`
```

If rsyslog is installed, you'll see something similar to the following command, with line 2 displaying syslog process information.

```

> 1. $ ps -ef |grep syslog

Reviewer 2: ```suggestion
1. $ ps -ef | grep syslog
```

> 1. $ ps -ef |grep syslog
> 2. 501 12734  3242   0  09:00AM ttys000    0:00.01 grep syslog

Reviewer 2: ```suggestion
$ ps -ef |grep syslog
501 12734  3242   0  09:00AM ttys000    0:00.01 grep syslog
```

```

If nothing is returned by this command, rsyslog is not installed.

If rsyslog is not installed, for installation instructions, please refer to the [Official Rsyslog Installation Documentation](https://www.rsyslog.com/doc/master/installation/index.html).

### Debugging

> If you run into issues, you may wish to debug your logs. This section will help guide you through that process for rsyslog using [environment variables](#Environment-Variables) or [command line](#Command-Line) to enable [`DebuggingOnDemand`](https://www.rsyslog.com/doc/v8-stable/troubleshooting/howtodebug.html#debug-on-demand).

Reviewer 3: ```suggestion
If you run into issues, you may wish to debug your logs. This section will help guide you through that process for rsyslog using [environment variables](#Environment-Variables) or [command line](#Command-Line) to enable [`DebugOnDemand`](https://www.rsyslog.com/doc/v8-stable/troubleshooting/howtodebug.html#debug-on-demand).
```

#### Environment Variables

To enable debugging via environment variables, add the following lines to your `rsyslog.conf` file, replacing the example path with a path of your own:

```

> 1. $DebugFile /somepath/example.log
> 2. $DebugLevel 1

Reviewer 2: ```suggestion
$DebugFile /somepath/example.log
$DebugLevel 1
```

```

A `DebugLevel` of two activates full debug mode. A level of 0 turns debug off, and a level of 1 activates `DebugOnDemand`.

#### Command Line

To enable debugging via your command line, open a terminal with root access to your application and execute the following command:

`export RSYSLOG_DEBUG="DebugOnDemand NoStdOut"`

next you need to specify where you want the debug logs to go to, you can do this by executing a command like below, replacing the example path with a path of your own:

`export RSYSLOG_DEBUGLOG=/somepath/example.log`

#### Restart rsyslog

Now that we've enabled debugging on demand and specified a location we want the debug logs to go to, we need to re-start rsyslog as a foreground application.

First, let's stop the rsyslog process by running the below terminal command:

`/etc/rc.d/init.d/rsyslog stop`

and then restart it again as a foreground process

`rsyslogd -n`

> Now we've configured `DebugOnDemmand` we can activate it (and deactivate it) on demand. To do this, in a new terminal window simply execute the following command:

Reviewer 3: typo
```suggestion
Now we've configured `DebugOnDemand` we can activate it (and deactivate it) on demand. To do this, in a new terminal window simply execute the following command:
```

> `` kill -USR1 `cat /var/run/rsyslogd.pid ``

Reviewer 2: ```suggestion
`` kill -USR1 `cat /var/run/rsyslogd.pid` ``
```
I think a backtick is missing here?

It could also use a sub command like so:

```suggestion
`kill -USR1 $(cat /var/run/rsyslogd.pid)`
```

You can now review logs in your debug log file.

<Notice>
  🔎 For further information on troubleshooting and debugging with syslog please
  refer to the [Official Syslog Trouble Shooting
  Documentation](https://www.rsyslog.com/doc/v8-stable/troubleshooting/).
</Notice>

## Retention

> Log Retention will be dependent on what logging plan you have with AppSignal. More information on pricing and retention will be available when Logging is no longer in Beta.

Reviewer 3: > logging plan

Are there going to be _logging plans_ separate from our regular plans?

Author: I believe so, but let's omit any suggestion to that until the Beta is over and the pricing is clear - and just keep the docs linking. Will remove this reference.

Beta Tester logs are retained for 30 days with a max capacity of 50GB.

> All information on publically available AppSignal plans, including storage and pricing, can be found on our [Pricing Page](https://www.appsignal.com/plans).

Reviewer 3: ```suggestion
All information on publicly available AppSignal plans, including storage and pricing, can be found on our [Pricing Page](https://www.appsignal.com/plans).
```
